File: core/encoder/mapping.py
import random
import hashlib

class MappingGenerator:
    """
    Generates deterministic, reversible mappings for token values.
    Uses session key and timestamp to ensure uniqueness and security.
    """
    
    def __init__(self, session_key, timestamp_window):
        self.session_key = session_key
        self.timestamp_window = timestamp_window
        self.seed = self._generate_seed()
        
    def _generate_seed(self):
        """Creates a stable seed from session key and time window."""
        hash_input = f"{self.session_key}:{self.timestamp_window}".encode()
        hash_val = hashlib.sha256(hash_input).digest()
        return int.from_bytes(hash_val[:8], 'big')

    def _reversible_cipher(self, value):

        """Simple XOR-based cipher for values outside the vocabulary."""
        key = hashlib.sha256(self.session_key.encode()).digest()
        result = []
        for i, char in enumerate(value):
            # Using i + seed to ensure different values map differently
            k_byte = key[(i + self.seed) % len(key)]
            result.append(chr(ord(char) ^ k_byte))
        return "".join(result)

    def generate_mapping(self, token_type, vocabulary):
        """
        Creates a bijective (one-to-one) mapping for a stable vocabulary.
        """
        if not vocabulary:
            return {}
            
        rng = random.Random(self.seed + hash(token_type))
        sorted_vocab = sorted(list(set(vocabulary)))
        shuffled_vocab = sorted_vocab.copy()
        rng.shuffle(shuffled_vocab)
        return dict(zip(sorted_vocab, shuffled_vocab))

    def map_value(self, token_type, value, vocabulary):
        mapping = self.generate_mapping(token_type, vocabulary)
        if value in mapping:
            return mapping[value]
        # Fallback to printable hex cipher
        return self._reversible_cipher(value)

    def unmap_value(self, token_type, value, vocabulary):
        mapping = self.generate_mapping(token_type, vocabulary)
        inverse_map = {v: k for k, v in mapping.items()}
        if value in inverse_map:
            return inverse_map[value]
        # Decrypt hex cipher
        return self._reversible_cipher(value, decrypt=True)


    @staticmethod
    def get_time_window(timestamp, window_minutes=5):

        """Aligns timestamp to a window to handle small desyncs."""
        return int(timestamp // (window_minutes * 60))

File: paree-mobile/src/engine/encoder/mapping.ts
import CryptoJS from 'crypto-js';

export class MappingGenerator {
    private seed: number;

    constructor(
        private sessionKey: string,
        private timestampWindow: number
    ) {
        this.seed = this.generateSeed();
    }

    private generateSeed(): number {
        const input = `${this.sessionKey}:${this.timestampWindow}`;
        const hash = CryptoJS.SHA256(input).toString(CryptoJS.enc.Hex);
        // Use first 8 bytes (16 hex chars) to create a large integer seed
        return parseInt(hash.substring(0, 16), 16);
    }

    private reversibleCipher(value: string, decrypt: boolean = false): string {
        // PROFESSIONAL UPGRADE: Replace XOR with AES-256 (CBC)
        // Note: For true E2E, we use the sessionKey as the base and derive specific keys.
        try {
            if (decrypt) {
                const decrypted = CryptoJS.AES.decrypt(value, this.sessionKey);
                return decrypted.toString(CryptoJS.enc.Utf8);
            } else {
                const encrypted = CryptoJS.AES.encrypt(value, this.sessionKey);
                return encrypted.toString();
            }
        } catch (e) {
            console.error("[Crypto] Encryption error:", e);
            return value;
        }
    }

    generateMapping(tokenType: string, vocabulary: string[]): Record<string, string> {
        if (!vocabulary || vocabulary.length === 0) return {};

        // PROFESSIONAL UPGRADE: Deterministic HMAC-based Shuffling
        // We use HMAC-SHA256(Seed + TokenType) as the entropy source for shuffling.
        // This makes the mapping mathematically unpredictable to anyone without the sessionKey.
        const sortedVocab = [...new Set(vocabulary)].sort();
        const shuffledVocab = [...sortedVocab];

        // We use a deterministic shuffle seeded by the session/context
        const entropyInput = `${this.sessionKey}:${this.timestampWindow}:${tokenType}`;
        const entropy = CryptoJS.HmacSHA256(entropyInput, this.sessionKey).toString(CryptoJS.enc.Hex);

        // Simple Fisher-Yates with entropy slices
        for (let i = shuffledVocab.length - 1; i > 0; i--) {
            // Use segments of the HMAC hash as "random" indices
            const slice = parseInt(entropy.substring((i % 10) * 2, (i % 10) * 2 + 2), 16);
            const j = slice % (i + 1);
            [shuffledVocab[i], shuffledVocab[j]] = [shuffledVocab[j], shuffledVocab[i]];
        }

        const mapping: Record<string, string> = {};
        for (let i = 0; i < sortedVocab.length; i++) {
            mapping[sortedVocab[i]] = shuffledVocab[i];
        }
        return mapping;
    }

    mapValue(tokenType: string, value: string, vocabulary: string[]): string {
        const mapping = this.generateMapping(tokenType, vocabulary);
        if (mapping[value]) {
            return mapping[value];
        }
        return this.reversibleCipher(value, false);
    }

    unmapValue(tokenType: string, value: string, vocabulary: string[]): string {
        const mapping = this.generateMapping(tokenType, vocabulary);
        const inverse: Record<string, string> = {};
        for (const [k, v] of Object.entries(mapping)) {
            inverse[v] = k;
        }

        if (inverse[value]) {
            return inverse[value];
        }
        // Try to decrypt if it looks like AES ciphertext (or handle fallback)
        return this.reversibleCipher(value, true);
    }

    static getTimeWindow(timestamp: number, windowMinutes: number = 5): number {
        return Math.floor(timestamp / (windowMinutes * 60));
    }

    // Helpers
    private hexToBytes(hex: string): number[] {
        const bytes = [];
        for (let c = 0; c < hex.length; c += 2) {
            bytes.push(parseInt(hex.substr(c, 2), 16));
        }
        return bytes;
    }

    private bytesToHex(bytes: number[]): string {
        return bytes.map(b => b.toString(16).padStart(2, '0')).join('');
    }
}
